lib/pacemaker/pcmk_output.c

```diff
@@ -582,6 +582,21 @@ stacks_and_constraints_xml(pcmk__output_t *out, va_list args) {
     return pcmk_rc_ok;
 }
 
+PCMK__OUTPUT_ARGS("health", "const char *", "const char *", "const char *", "const char *")
+static int
+health_default(pcmk__output_t *out, va_list args)
```

Comment:
Looking good. Do you think we'll still need the new default functions when this is done? Making sure `pcmk__formatted_printf()` is a no-op for non-text ought to cover it, if the rest of the code is the same.

If we're worried about the semantic issue of calling that function at all in a default formatter, we could guard the call.

Reply:
Hm, maybe not.  I'm torn on how to proceed here.

On the one hand, I just dislike calling pcmk__formatted_printf from a non-text message handler.  That function isn't meant to be used in such a way and I don't like leaving the calls there as an example to be copied.  We also don't do very much with ->is_quiet outside of the text handlers.  Removing that stuff in the default handlers does result in some duplication, but the formatted output code typically chooses being straighforward and redundant over reuse.

We typically do not guard calls by checking format name in message handlers unless there is absolutely no choice, so I'd hate to start doing that.

On the other hand, you're right that as long as is_quiet is false or as long as pcmk__formatted_printf does nothing, then these functions are the same as the defaults.  So I guess here my concerns really come down more to style than anything.

Reply:
I guess it is nice to have something that will be used for "log" by default.

To reduce duplication maybe the text implementations could call the default implementations when appropriate.

Reply:
> To reduce duplication maybe the text implementations could call the default implementations when appropriate.

Is there any reason why we might want to avoid adding checks on ->fmt_name in these functions to accomplish this?

Reply:
> I just dislike calling pcmk__formatted_printf from a non-text message handler.

That's one point in favor of some type of non-newline `printf` method for each class of output object. The obvious point against it is that it only makes sense for text output so it's kind of silly to have one (even a no-op) for every class. It may look cleaner at the caller level, idk

Reply:
I suppose that's a good reason to add ->printf (or whatever).  That makes it impossible for the wrong formatter to call the function.  On the other hand, there's already four printing functions defined in lib/common/output_text.c - pcmk__formatted_(v)printf and pcmk__indented(v)printf.  So it could end up being a lot of things added to the format struct.

Reply:
> > To reduce duplication maybe the text implementations could call the default implementations when appropriate.
> 
> Is there any reason why we might want to avoid adding checks on ->fmt_name in these functions to accomplish this?

Not sure why we would need to ... the default methods shouldn't check fmt_name since by definition they should work with arbitrary formats.

> I suppose that's a good reason to add ->printf (or whatever). That makes it impossible for the wrong formatter to call the function. On the other hand, there's already four printing functions defined in lib/common/output_text.c - pcmk__formatted_(v)printf and pcmk__indented(v)printf. So it could end up being a lot of things added to the format struct.

The methods should be semantic, and not indicate something that should be done a certain way. Like the difference between emphasis and italics.

The *printf() functions are basically info() ignoring quiet. So maybe we need a notice() method that gets displayed regardless of quiet.